File: runners/nodejs/src/interfaces/callback_channel.ts
import { Writer } from "./writer";
import { Channel } from "./channel";

/**
 * A callback channel is a simple implementation of a writer that calls a
 * callback whenever a value is written to it. The class does therefore
 * not implement the `Reader` interface, as it is not possible to read from
 * the channel.
 */
export class CallbackChannel<T> implements Writer<T> {
  /**
   * The flow of messages is implemented by wrapping a regular channel.
   * @private
   */
  private channel = new Channel<T>();

  /**
   * The callback that is called whenever a value is written to the channel.
   * @private
   */
  private readonly onWrite: (value: T) => Promise<void>;

  /**
   * A callback to execute when the channel is closed. Note that this will wait
   * for the remaining data to be written before closing the channel.
   * @private
   */
  private readonly onClose: null | (() => Promise<void>) = null;

  /**
   * Handle incoming messages of the channel by piping them into the callback.
   * @private
   */
  private handler = new Promise(async () => {
    for await (const data of this.channel) {
      await this.onWrite(data)
    }

    if (this.onClose != null) {
      await this.onClose()
    }
  });

  /**
   * Create a new callback channel with a specific callback.
   * @param onWrite The callback to call whenever a value is written to the
   * @param onClose An optional callback to execute when the channel is closed.
   * channel.
   */
  constructor(
    onWrite: (value: T) => Promise<void>,
    onClose: (() => Promise<void>) | null = null,
  ) {
    this.onWrite = onWrite;
    this.onClose = onClose;
  }

  close(): Promise<void> {
    this.channel.close()
    return Promise.resolve();
  }

  isClosed(): boolean {
    return this.channel.isClosed();
  }

  write(data: T) {
    this.channel.write(data)
  }
}

File: packages/runner-py/src/rdfc/util/callback_channel.py
import rdfc
import typing

OnWrite = typing.Callable[[bytes], typing.Coroutine[typing.Any, typing.Any, None]]
OnClose = typing.Callable[[], typing.Coroutine[typing.Any, typing.Any, None]]


class CallbackChannel(rdfc.Writer):
    closed: bool = False
    on_write: OnWrite
    on_close: OnClose

    def __init__(
        self,
        on_write: OnWrite,
        on_close: OnClose,
    ):
        self.on_write = on_write
        self.on_close = on_close

    async def write(self, value: bytes) -> None:
        await self.on_write(value)

    async def close(self) -> None:
        if self.closed is False:
            self.closed = True
            await self.on_close()

    def is_closed(self) -> bool:
        return self.closed
